Automattic/sensei
# Add a course setting to not allow self-enrollment

Resolves #7182

## Proposed Changes

* Add a course setting to not allow self enrollment.
* Also add notices in the frontend when the setting is enabled.

## Testing Instructions


1. Create a course with blocks and Learning Mode enabled.
2. Enable the setting _"Don't allow self enrollment"_ in the course editor sidebar.
3. Add at least one lesson to the course.
4. As a student, visit the course and the lesson pages.
5. Make sure you see notices and you can't enroll.
6. Enroll the student through the Student Management (WP Admin), and make sure it works properly.
7. Check that you don't see the notices anymore.
8. Repeat the previous steps, but this time with a course without blocks (legacy course), and without Learning Mode.
9. Play around, and make sure there's not take course button for the user when the new setting is enabled.
10. Turn the setting _"Don't allow self enrollment"_ off, and make sure the course works normally (no notices, and it's possible to enroll).

### Known issues

https://github.com/Automattic/sensei/issues/7232

### Screenshots

The setting:
<img width="315" alt="Screenshot 2023-10-18 at 11 40 34" src="https://github.com/Automattic/sensei/assets/876340/cbaed326-1158-4da3-84f4-83655cb01b9a">

Notice on course with blocks:
<img width="1112" alt="Screenshot 2023-10-18 at 14 58 44" src="https://github.com/Automattic/sensei/assets/876340/21c1cfa0-c5f1-44d2-9bc5-a2292d15a3d2">

Notice on lesson with Learning Mode:
<img width="1203" alt="Screenshot 2023-10-18 at 11 41 02" src="https://github.com/Automattic/sensei/assets/876340/acf68e71-aeea-4d84-9998-d03746ca66f9">

Notice on course without blocks (impacted by https://github.com/Automattic/sensei/issues/6173 in some themes):
<img width="1460" alt="Screenshot 2023-10-18 at 12 16 14" src="https://github.com/Automattic/sensei/assets/876340/3fc8fe1e-8a7b-4b9d-83a1-cfecfee1d181">

Notice on lesson without Learning Mode:
<img width="1119" alt="Screenshot 2023-10-17 at 16 42 50" src="https://github.com/Automattic/sensei/assets/876340/19d2e115-afe3-4c74-8499-85012de0c6c0">

Courses list (impacted by https://github.com/Automattic/sensei/issues/7232):
<img width="1192" alt="Screenshot 2023-10-18 at 16 28 17" src="https://github.com/Automattic/sensei/assets/876340/79304ace-f90d-4597-9516-f058f51c00c2">

## New/Updated Hooks


* `sensei_self_enrollment_not_allowed` - Filter whether self enrollment is not allowed for a given course.

## Pre-Merge Checklist
